### PR TITLE
REVIEW AND MERGE: the addition of sklearn mirror of knn_lrr for validation

### LINKED ISSUE #109: Implement sklearn mirror of from-scratch KNN + Logistic Ridge Regression for validation (Story Points: 3)
## Description

Build a sklearn-based mirror of our from-scratch `knn_lrr` algorithm (KNN with locally-fitted Logistic Ridge Regression) to serve as a reference implementation for validation. Given the same query, crime type, and `k`, both implementations should produce numerically close arrest probability predictions. The from-scratch version remains the academic deliverable; this sklearn version is the trusted benchmark we compare against.

## Tasks

- [ ] Create new module (e.g. `algorithms/knn_sklearn.py`) with function `knn_sklearn(query, crime_type=None, k=None)`
- [ ] Reuse `select_crime` (or refactor into shared helper) to filter dataset by `primary_type`
- [ ] Use the same feature set: `latitude`, `longitude`, and the 8 cyclical time features (`hour_sin/cos`, `day_of_week_sin/cos`, `month_sin/cos`, `day_of_year_sin/cos`); label is `arrest`
- [ ] Standardize features with `StandardScaler` (fit on filtered dataset, apply same transform to query)
- [ ] Find k nearest neighbors of the query using `sklearn.neighbors.NearestNeighbors(algorithm='kd_tree', metric='euclidean')`
- [ ] Fit `sklearn.linear_model.LogisticRegression(penalty='l2', C=1.0, solver='lbfgs')` locally on those k neighbors
- [ ] Confirm and document the `C` ↔ `alpha` mapping in a code comment (from-scratch uses `alpha=1.0`, which corresponds to `C=1.0`)
- [ ] Do **not** manually augment features with an intercept column — let sklearn handle the intercept (`fit_intercept=True` by default, unpenalized)
- [ ] Predict arrest probability for the query and return it
- [ ] Mirror the diagnostic print output (intercept, coefficients, probability, scenario message) for side-by-side comparison
- [ ] Write validation script/notebook cell running both implementations on the same inputs across at least 3 crime types, printing predicted probability, top-5 coefficients, and neighbor index overlap
- [ ] Verify predictions agree within tolerance (|Δp| < 0.05) on a fixed test query
- [ ] Set `random_state` wherever applicable for reproducibility
- [ ] Add docstring + brief markdown cell in the relevant notebook explaining this is a validation mirror

## Dependencies

- From-scratch `knn` implementation (already merged)
- `select_crime` helper and `preprocess_data` pipeline (already functional)
- `data/cleaned/chicago_crimes_cleaned.parquet` available locally
- `scikit-learn` added to project dependencies (verify it's in `pyproject.toml` / `requirements.txt`)
- No blocker on dashboard work — this is standalone validation tooling

### PR DESCRIPTION
**Pull request by carl-harper**
## What does this PR do?
Adds algorithms/knn_sklearn.py — a sklearn-based mirror of the from-scratch knn_lrr algorithm. Uses NearestNeighbors + locally-fitted LogisticRegression(penalty='l2') on the same feature set (latitude, longitude, 8 cyclic time features) to serve as a trusted benchmark for validating the from-scratch implementation. Predictions should agree within |Δp| < 0.05.

Pull request specifically: review and merge if it works

## How can a reviewer test this?

from algorithms.knn_sklearn import knn_sklearn

test_query = {
    "latitude": 41.8827, "longitude": -87.6233,
    "hour_sin": 0.0, "hour_cos": -1.0,
    "day_of_week_sin": 0.782, "day_of_week_cos": 0.623,
    "month_sin": 0.0, "month_cos": -1.0,
    "day_of_year_sin": 0.995, "day_of_year_cos": -0.100,
}
prob = knn_sklearn(query=test_query, crime_type="THEFT", k=100)
print(prob)  # should print a float between 0 and 1

## Any edge cases to be aware of?

If all k neighbours share the same arrest label, the function returns the mean directly (LogisticRegression requires both classes present)
crime_type must match primary_type spelling exactly in the dataset
Validation script (validate_knn_sklearn.py) is not yet committed — that comes in a follow-up once knn_lrr's call signature is confirmed

## Checklist

 Tests written
 No direct push to main
 Reviewed by at least one teammate

### CODE REVIEW

**Comment by wangrgrace** [comment]
The code works fine. If we end up implementing it into the Algorithm section of the Flask app, just make sure the numbers make sense.
